### PR TITLE
EdgeHub Restart Tester (Sender Refactored)

### PR DESCRIPTION
**Pull request by yophilav**
1. use from ReportTestResultUntilSuccessAsync() to be`ReportTestResultAsync()
2. Update the DM and Msg connector such that its constructor create the moduleClient in the connector constructor, and the StartAsync(eachTestExpirationTime, restartTime, cts.Token) in the while loop to do the send. 
3. Remove the lock increment/decrement for msgCount and dmCount

### CODE REVIEW

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs [hunk 1]
```diff
@@ -18,60 +18,67 @@ namespace EdgeHubRestartTester
...
-        uint restartSequenceNumber;
         ILogger logger;
 
+        ModuleClient DmModuleClient
+        {
+            get
+            {
+                if (this.dmModuleClient == null)
+                {
+                    this.dmModuleClient = ModuleUtil.CreateModuleClientAsync(
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs]: need "await"; make it as a method and call it in StartAsync method.

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs [hunk 2]
```diff
@@ -18,60 +18,67 @@ namespace EdgeHubRestartTester
...
-            this.cancellationToken = Preconditions.CheckNotNull(cancellationToken, nameof(cancellationToken));
-            this.edgeHubRestartedTime = edgeHubRestartedTime;
-            this.restartSequenceNumber = restartSequenceNumber;
             this.logger = Preconditions.CheckNotNull(logger, nameof(logger));
         }
 
-        public async Task StartAsync()
+        public void Dispose() => this.dmModuleClient?.Dispose();
+
+        public async Task StartAsync(
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs]: Need restart count as argument

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs [hunk 3]
```diff
@@ -18,60 +18,67 @@ namespace EdgeHubRestartTester
...
                 this.logger).ConfigureAwait(false);
 
             TestResultBase dmTestResult = new EdgeHubRestartDirectMethodResult(
                 Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartDirectMethod.ToString(),
                 DateTime.UtcNow,
                 Settings.Current.TrackingId,
                 this.batchId,
-                Interlocked.Read(ref this.directMethodCount).ToString(),
-                this.edgeHubRestartedTime,
+                this.directMethodCount.ToString(),
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs]: what is the expected value of direct method count to report to TRC for each run?

File: test/modules/EdgeHubRestartTester/src/IEdgeHubConnector.cs [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
+    using System;
+    using System.Threading;
     using System.Threading.Tasks;
 
-    public interface IEdgeHubConnector
+    public interface IEdgeHubConnector : IDisposable
     {
-        Task StartAsync();
+        Task StartAsync(
+            DateTime runExpirationTime,
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/IEdgeHubConnector.cs]: Add restartCount argument

File: test/modules/EdgeHubRestartTester/src/Program.cs [hunk 5]
```diff
@@ -72,31 +65,20 @@ static async Task<int> MainAsync()
                     Task sendMessageTask = Task.CompletedTask;
                     if (Settings.Current.MessageEnabled)
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/Program.cs]:
```suggestion
                    if (edgeHubMessageConnector != null)
```

File: test/modules/EdgeHubRestartTester/src/IEdgeHubConnector.cs (hunk 4 above)
Comment (yophilav) [reply test/modules/EdgeHubRestartTester/src/IEdgeHubConnector.cs]: I don't need this argument anymore.

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs (hunk 3 above)
Comment (yophilav) [reply test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnector.cs]: Increasing number ( not necessary incremental ).

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs [hunk 6]
```diff
@@ -17,61 +17,51 @@ namespace EdgeHubRestartTester
 
     class DirectMethodEdgeHubConnectorTest : IEdgeHubConnector
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]:
```suggestion
    class DirectMethodEdgeHubConnectorTest : IEdgeHubConnectorTest
```

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs [hunk 7]
```diff
@@ -88,14 +78,14 @@ public async Task StartAsync()
                 try
                 {
                     // Direct Method sequence number is always increasing regardless of sending result.
-                    Interlocked.Increment(ref this.directMethodCount);
+                    this.directMethodCount++;
                     MethodRequest request = new MethodRequest(
                         directMethodName,
-                        Encoding.UTF8.GetBytes($"{{ \"Message\": \"Hello\", \"DirectMethodCount\": \"{Interlocked.Read(ref this.directMethodCount).ToString()}\" }}"),
+                        Encoding.UTF8.GetBytes($"{{ \"Message\": \"Hello\", \"DirectMethodCount\": \"{this.directMethodCount.ToString()}\" }}"),
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]:
```suggestion
                        Encoding.UTF8.GetBytes($"{{ \"Message\": \"Hello\", \"DirectMethodCount\": \"{this.directMethodCount}\" }}"),
```

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs [hunk 8]
```diff
@@ -17,61 +17,51 @@ namespace EdgeHubRestartTester
...
+        public async Task StartAsync(
+            DateTime runExpirationTime,
+            DateTime edgeHubRestartedTime,
+            CancellationToken cancellationToken)
         {
             (DateTime dmCompletedTime, HttpStatusCode dmStatusCode) = await this.SendDirectMethodAsync(
                 Settings.Current.DeviceId,
                 Settings.Current.DirectMethodTargetModuleId,
-                this.dmModuleClient,
+                await this.GetModuleClientAsync(),
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]: call GetModuleClientAsync method inside SendDirectMethodAsync.

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs [hunk 9]
```diff
@@ -17,61 +17,51 @@ namespace EdgeHubRestartTester
...
                 Settings.Current.DirectMethodName,
-                this.runExpirationTime,
-                this.cancellationToken,
+                runExpirationTime,
+                cancellationToken,
                 this.logger);
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]: don't need to pass this.logger to this method.

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs [hunk 10]
```diff
@@ -114,20 +104,58 @@ public async Task StartAsync()
...
+        {
+            if (this.reportClient == null)
+            {
+                this.reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
+            }
+
+            return this.reportClient;
+        }
+
+        string GetSourceString() => Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartDirectMethod.ToString();
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]: nit: use string interpolation

File: test/modules/EdgeHubRestartTester/src/IEdgeHubConnector.cs [hunk 11]
```diff
@@ -1,10 +1,15 @@
 // Copyright (c) Microsoft. All rights reserved.
 namespace EdgeHubRestartTester
 {
+    using System;
+    using System.Threading;
     using System.Threading.Tasks;
 
-    public interface IEdgeHubConnector
+    public interface IEdgeHubConnector : IDisposable
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/IEdgeHubConnector.cs]:
```suggestion
    public interface IEdgeHubConnectorTest : IDisposable
```

File: test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnectorTest.cs [hunk 12]
```diff
@@ -14,60 +14,82 @@ namespace EdgeHubRestartTester
...
+        {
+            if (this.reportClient == null)
+            {
+                this.reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
+            }
+
+            return this.reportClient;
+        }
+
+        string GetSourceString() => Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartMessage.ToString();
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnectorTest.cs]:
```suggestion
        string GetSource() => Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartMessage.ToString();
```

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs (hunk 10 above)
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]:
```suggestion
        string GetSource() => Settings.Current.ModuleId + "." + TestOperationResultType.EdgeHubRestartDirectMethod.ToString();
```

File: test/modules/EdgeHubRestartTester/src/Program.cs [hunk 13]
```diff
@@ -62,42 +56,27 @@ static async Task<int> MainAsync()
...
                 while ((!cts.IsCancellationRequested) && (DateTime.UtcNow < testExpirationTime))
                 {
-                    (DateTime restartTime, _) = await RestartEdgeHub(iotHubServiceClient);
+                    (DateTime restartTime, _) = await RestartEdgeHub(
+                        iotHubServiceClient,
+                        cts.Token);
                     DateTime eachTestExpirationTime = restartTime.Add(Settings.Current.RestartPeriod);
 
                     // Increment the counter when issue an edgeHub restart
                     restartCount++;
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/Program.cs]: where to use? do you still need this?

File: test/modules/EdgeHubRestartTester/src/Program.cs [hunk 14]
```diff
@@ -126,7 +105,8 @@ static async Task<int> MainAsync()
         }
 
         static async Task<Tuple<DateTime, HttpStatusCode>> RestartEdgeHub(
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/Program.cs]: Please update this method to use docker or any other approach to support offline scenario before working on report generation.

File: test/modules/EdgeHubRestartTester/src/Program.cs [hunk 15]
```diff
@@ -161,12 +141,29 @@ static async Task<int> MainAsync()
 
                     if (restartCount == maxRetry - 1)
                     {
-                        throw;
+                        string errorMessage = $"Failed to restart EdgeHub with payload: {payload}: {e}";
+                        TestResultBase errorResult = new ErrorTestResult(
+                            Settings.Current.TrackingId,
+                            GetSourceString(),
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/Program.cs]: use module id should be good enough

File: test/modules/ModuleLib/TestResults/EdgeHubRestartMessageResult.cs [hunk 16]
```diff
@@ -27,7 +27,7 @@ public class EdgeHubRestartMessageResult : MessageTestResult
             this.MessageCompletedStatusCode = messageCompletedStatusCode;
         }
 
-        DateTime EdgeHubRestartedTime { get; set; }
+        public DateTime EdgeHubRestartedTime { get; set; }
 
         public DateTime MessageCompletedTime { get; set; }
```
Comment (philipktlin) [inline test/modules/ModuleLib/TestResults/EdgeHubRestartMessageResult.cs]: Why do you need "GetMessageTestResult()"?

File: test/modules/EdgeHubRestartTester/src/Program.cs (hunk 13 above)
Comment (yophilav) [reply test/modules/EdgeHubRestartTester/src/Program.cs]: Oh this should be removed.

File: test/modules/EdgeHubRestartTester/src/Program.cs (hunk 14 above)
Comment (yophilav) [reply test/modules/EdgeHubRestartTester/src/Program.cs]: This is a bigger task than I want to tackle right now. 
Would it be alright I if I get all the refactoring + verification parth in first? then revisit the restart?

File: test/modules/ModuleLib/TestResults/EdgeHubRestartMessageResult.cs (hunk 16 above)
Comment (yophilav) [reply test/modules/ModuleLib/TestResults/EdgeHubRestartMessageResult.cs]: It's used to extract the `batchId;seqNum;trackId`

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs [hunk 17]
```diff
@@ -15,95 +15,82 @@ namespace EdgeHubRestartTester
...
                 dmCompletedTime,
-                dmStatusCode,
-                this.restartSequenceNumber);
+                dmStatusCode);
 
-            var reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
-            await ModuleUtil.ReportTestResultUntilSuccessAsync(
-                reportClient,
+            await ModuleUtil.ReportTestResultAsync(
+                this.GetReportClient(),
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]: remove this.GetReportClient() and this.logger from this method call.

File: test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnectorTest.cs [hunk 18]
```diff
@@ -14,81 +14,103 @@ namespace EdgeHubRestartTester
...
+                this.messageCount.ToString(),
+                edgeHubRestartedTime,
                 msgCompletedTime,
                 mgsStatusCode);
 
-            var reportClient = new TestResultReportingClient { BaseUrl = Settings.Current.ReportingEndpointUrl.AbsoluteUri };
-            await ModuleUtil.ReportTestResultUntilSuccessAsync(
-                reportClient,
+            await ModuleUtil.ReportTestResultAsync(
+                this.GetReportClient(),
```
Comment (philipktlin) [inline test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnectorTest.cs]: remove this.GetReportClient() and this.logger here and move into the method.

File: test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs (hunk 17 above)
Comment (yophilav) [reply test/modules/EdgeHubRestartTester/src/DirectMethodEdgeHubConnectorTest.cs]: `ReportTestResultAsync()` lives in `ModuleUtil` class which shared across multiple classes, not only within `IEdgeHubConnectorTest`

File: test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnectorTest.cs (hunk 18 above)
Comment (yophilav) [reply test/modules/EdgeHubRestartTester/src/MessageEdgeHubConnectorTest.cs]: `ReportTestResultAsync()` lives in `ModuleUtil` class which shared across multiple classes, not only within `IEdgeHubConnectorTest`

**Pull request merged**